### PR TITLE
refactor(common): use a specialized curry alternative for validation

### PR DESCRIPTION
**Pull request by kszucs**
With a slightly less convenient rules API we should achieve some speedup for argument validation, still need to write benchmarks though.

### CODE REVIEW

File: ibis/common/validators.py [hunk 1]
```diff
@@ -2,20 +2,18 @@
...
 
 EMPTY = inspect.Parameter.empty  # marker for missing argument
 
 
-class Validator(Callable):
-    """
-    Abstract base class for defining argument validators.
-    """
+class Validator:
+    __slots__ = tuple()
```
Comment (cpcloud) [inline ibis/common/validators.py:13]:
```suggestion
    __slots__ = ()
```

File: ibis/common/validators.py [hunk 2]
```diff
@@ -30,13 +28,19 @@ class Optional(Validator):
...
+        'default',
+    )
 
     def __init__(self, validator, default=None):
         self.validator = validator
         self.default = default
 
-    def __call__(self, arg, **kwargs):
+    def __repr__(self):
+        return f"Optional({self.validator!r}, {self.default!r})"
```
Comment (cpcloud) [inline ibis/common/validators.py:41]:
```suggestion
        return f"{self.__class__.__name__}({self.validator!r}, {self.default!r})"
```

File: ibis/common/validators.py [hunk 3]
```diff
@@ -45,23 +49,35 @@ def __call__(self, arg, **kwargs):
...
+        new_kwargs = {**self.kwargs, **kwargs}
+        return self.__class__(self.func, new_args, new_kwargs)
 
     def __repr__(self):
-        return '{}({}{})'.format(
-            self.func.__name__,
-            repr(self.args)[1:-1],
-            ', '.join(f'{k}={v!r}' for k, v in self.keywords.items()),
+        return (
+            f"CurriedValidator({self.func!r}, {self.args!r}, {self.kwargs!r})"
```
Comment (cpcloud) [inline ibis/common/validators.py:72]:
```suggestion
            f"{self.__class__.__name__}({self.func!r}, {self.args!r}, {self.kwargs!r})"
```

File: ibis/common/validators.py [hunk 4]
```diff
@@ -45,23 +49,35 @@ def __call__(self, arg, **kwargs):
...
-    Enable convenient validator definition by decorating plain functions.
-    """
+class CurriedValidator(Validator):
+
+    __slots__ = ("func", "args", "kwargs")
+
+    def __init__(self, func, args=None, kwargs=None):
+        # TODO(kszucs): validate that func is callable
+        self.func = func
+        self.args = args or tuple()
```
Comment (cpcloud) [inline ibis/common/validators.py:62]:
```suggestion
        self.args = args or ()
```

File: ibis/common/validators.py [hunk 5]
```diff
@@ -30,13 +28,19 @@ class Optional(Validator):
...
 
     def __init__(self, validator, default=None):
         self.validator = validator
         self.default = default
 
-    def __call__(self, arg, **kwargs):
+    def __repr__(self):
+        return f"Optional({self.validator!r}, {self.default!r})"
+
+    def validate(self, arg, **kwargs):
```
Comment (cpcloud) [inline ibis/common/validators.py:43]: Why can't we preserve `__call__`?

Comment (kszucs) [reply ibis/common/validators.py:43]: It's an experimental PR to see the performance gain if any. 

Preserving `__call__` would require a decent amount of signature tracking, see the `toolz.curry` implementation: https://github.com/pytoolz/toolz/blob/f3ac8211f510edff77c1479061d99bb2dc8ab878/toolz/functoolz.py#L170-L377

**Comment by cpcloud** [comment]
@kszucs needs a rebase, then we can merge it.

**Comment by kszucs** [comment]
If there is no significant benchmark improvement then we should keep the earlier curry based implementation since that's more convenient.

**Comment by cpcloud** [comment]
> If there is no significant benchmark improvement then we should keep the earlier curry based implementation since that's more convenient.

Got it. It might be useful to see whether this PR fixes rules docstrings, which are currently not working for anything curried.

**Comment by cpcloud** [comment]
I futzed with this a bit to see if it helps with the `__doc__` issue, and it doesn't, so we can close it out.

**Pull request closed without merging**